rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_service.hpp

```diff
@@ -0,0 +1,145 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP_LIFECYCLE__LIFECYCLE_SERVICE_HPP_
+#define RCLCPP_LIFECYCLE__LIFECYCLE_SERVICE_HPP_
+
+#include <memory>
+#include <string>
+
+#include "rclcpp/any_service_callback.hpp"
+#include "rclcpp/service.hpp"
+
+namespace rclcpp_lifecycle
+{
+/// base class with only
+/**
+ * pure virtual functions. A managed
+ * node can then deactivate or activate
+ * the service handling.
+ * It is more a convenient interface class
+ * than a necessary base class.
+ */
+class LifecycleServiceInterface
+{
+public:
+  virtual ~LifecycleServiceInterface() {}
+  virtual void on_activate() = 0;
+  virtual void on_deactivate() = 0;
+  virtual bool is_activated() = 0;
+};
+
+/// brief child class of rclcpp Service class.
+/**
+ * Overrides all service functions to check for enabled/disabled state.
+ */
+template<typename ServiceT>
+class LifecycleService : public LifecycleServiceInterface, public rclcpp::Service<ServiceT>
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(LifecycleService)
+
+  /// Default constructor.
+  /**
+   * The constructor for a LifecycleService is almost never called directly.
+   * Instead, services should be instantiated through the function
+   * rclcpp::create_service().
+   *
+   * \param[in] node_handle NodeBaseInterface pointer that is used in part of
+   *   the setup.
+   * \param[in] service_name Name of the topic to publish to.
+   * \param[in] any_callback User defined callback to call when a client request
+   *   is received.
+   * \param[in] service_options options for the subscription.
+   */
+  LifecycleService(
```

Comment:
can we just add a using declaration of the parent class constructor?
i.e.:

```cpp
using rclcpp::Service<ServiceT>::Service;
```

It seems to me that we're adding exactly the same constructors here.
`LifecycleServiceInterface` and `enabled_` can be default initialized, so that should work.

Note: Maybe it's `using rclcpp::Service<ServiceT>::Service<ServiceT>;`, I always get confused with that.

Reply:
@ivanpauno Done.